src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -414,6 +434,18 @@ void FileSegment::write(const char * from, size_t size, size_t offset)
         e.addMessage(fmt::format("{}, current cache state: {}", e.what(), getInfoForLogUnlocked(lk)));
         setDownloadFailedUnlocked(lk);
 
+        if (external_wb)
+        {
+            try
+            {
+                external_wb->finalize();
```

Comment:
Why do we need `finalize` in case of error? Can't it be responsibility of the caller?